File: app/src/main/java/com/namoadigital/prj001/ui/act005/Act005_Opc.java
package com.namoadigital.prj001.ui.act005;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.SimpleAdapter;

import com.namoa_digital.namoa_library.util.HMAux;
import com.namoa_digital.namoa_library.util.ToolBox;
import com.namoadigital.prj001.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by DANIEL.LUCHE on 31/01/2017.
 */

public class Act005_Opc extends Fragment {

    public static final String DRAWER_KEY_OPC_ID = "drawer_opc_id" ;
    public static final String DRAWER_KEY_OPC_ICON = "drawer_opc_icon";
    public static final String DRAWER_KEY_OPC_DESC = "drawer_opc_desc" ;
    //
    public static final String DRAWER_OPC_CUSTOMER = "drawer_opc_customer" ;
    public static final String DRAWER_OPC_SITE = "drawer_opc_site" ;
    public static final String DRAWER_OPC_OPERATION = "drawer_opc_operation" ;
    public static final String DRAWER_OPC_LOGOUT = "drawer_opc_logout" ;
    //
    private ImageView iv_logo;
    private ListView lv_opc;
    private IAct005_Opc delegate;
    private List<HMAux> drawerItemList;
    private HMAux hmAux_Trans;
    private SimpleAdapter sAdapter;

    public interface IAct005_Opc{
        void itemClicked(String index);
    }

    public void setOnOpcItemClicked(IAct005_Opc delegate) {
        this.delegate = delegate;
    }
    //

    public void setHmAux_Trans(HMAux hmAux_Trans,String mModule_Code , String mResource_Code) {
        this.hmAux_Trans = hmAux_Trans;
        translateItens(mModule_Code, mResource_Code);
        sAdapter.notifyDataSetChanged();
    }

    private void translateItens(String mModule_Code, String mResource_Code) {
        for (HMAux item : drawerItemList) {
            if (hmAux_Trans.get(item.get(Act005_Opc.DRAWER_KEY_OPC_DESC)) != null) {
                item.put(Act005_Opc.DRAWER_KEY_OPC_DESC, hmAux_Trans.get(item.get(Act005_Opc.DRAWER_KEY_OPC_DESC)));
            } else {
                item.put(Act005_Opc.DRAWER_KEY_OPC_DESC, ToolBox.setNoTrans(mModule_Code, mResource_Code, item.get(Act005_Opc.DRAWER_KEY_OPC_DESC)));
            }
        }
    }


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.act005_opc_content,container,false);
        //
        iniVar(view);
        iniAction();
        //
        return view;

    }

    private void iniVar(View view) {
        iv_logo = (ImageView) view.findViewById(R.id.act005_opc_iv_logo);
        //
        lv_opc = (ListView) view.findViewById(R.id.act005_opc_lv_opt);
        //
        lvSetup();
    }

    private void lvSetup() {
        String[] from = {DRAWER_KEY_OPC_ICON, DRAWER_KEY_OPC_DESC};
        int[] to ={R.id.act005_opc_cell_iv_icon, R.id.act005_opc_cell_tv_desc};
        sAdapter = new SimpleAdapter(
                getActivity(),
                loadOptions(),
                R.layout.act005_opc_cell,
                from,
                to
        );

        lv_opc.setAdapter(sAdapter);

    }

    private void iniAction() {
        lv_opc.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                if(delegate != null){
                    HMAux item = (HMAux) parent.getItemAtPosition(position);
                    delegate.itemClicked(item.get(DRAWER_KEY_OPC_ID));
                }
            }
        });
    }

    private List<HMAux> loadOptions() {
        String[] id = {
                DRAWER_OPC_CUSTOMER,
                DRAWER_OPC_SITE,
                DRAWER_OPC_OPERATION,
                DRAWER_OPC_LOGOUT
        };
        String[] icon = {
                "",
                "",
                "",
                String.valueOf(R.drawable.ic_settings_power_red_24dp)
        };

        String[] desc = {
                "lbl_change_customer",
                "lbl_change_site",
                "lbl_change_operation",
                "lbl_logout",
        };
        drawerItemList = new ArrayList<>();
        for (int i = 0; i < id.length;i++){
            HMAux hmAux =  new HMAux();
            hmAux.put(DRAWER_KEY_OPC_ID,id[i]);
            hmAux.put(DRAWER_KEY_OPC_ICON, icon[i]);
            hmAux.put(DRAWER_KEY_OPC_DESC,desc[i]);
            drawerItemList.add(hmAux);
        }

        return drawerItemList;
    }
}

File: app/src/main/java/com/namoadigital/prj001/ui/act005/Act005Opc.kt
package com.namoadigital.prj001.ui.act005

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.Nullable
import androidx.fragment.app.Fragment
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.namoa_digital.namoa_library.util.HMAux
import com.namoadigital.prj001.databinding.Act005OpcContentBinding
import com.namoadigital.prj001.util.ConstantBaseApp

class Act005Opc : Fragment() {
    private lateinit var binding: Act005OpcContentBinding
    private lateinit var hmAux_Trans: HMAux
    private var delegate: Act005DrawerInteraction? = null
    private lateinit var logoReceiver: CustomerLogoReceiver

    /**
     * Interface com os metodo usados no frag e que deve ser implementado pela act host
     */
    interface Act005DrawerInteraction{
        @Nullable
        fun getCustomerLogo(): Bitmap
        fun hasPendencies() : Boolean
        fun showEnableNfcOption(): Boolean
        fun showDisableNfcOption(): Boolean
        fun showChangeCustomerOption(): Boolean
        fun onHistoricClick()
        fun onEnableNfcClick()
        fun onDisableNfcClick()
        fun onSupportDataClick()
        fun onChangeCustomerClick()
        fun onLogoutClick()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = Act005OpcContentBinding.inflate(inflater, container, false)
        iniVars()
        iniActions()
        return binding.root
    }

    /**
     * Seta interface no momento do attach ou for exception se não implementado
     */
    override fun onAttach(context: Context) {
        super.onAttach(context)
        if(context is Act005DrawerInteraction){
            delegate = context as Act005DrawerInteraction
        }else{
            throw Exception("Act005DrawerInteraction not implemented")
        }
    }

    /**
     * Remove a interface
     */
    override fun onDetach() {
        super.onDetach()
        delegate = null
    }

    private fun iniVars() {
        setLabels()
        setSetupViews()
        initializeLogoReceiver()
    }

    /**
     * Metodo que iniciaçiza o receiver do logo
     */
    private fun initializeLogoReceiver() {
        if(!this::logoReceiver.isInitialized){
            logoReceiver = CustomerLogoReceiver()
        }
    }

    /**
     * Metodo que seta hmAux na propriedade.
     */
    fun setHmAux_Trans(hmAux_Trans: HMAux){
        this.hmAux_Trans = hmAux_Trans
    }

    /**
     * Metodo que seta as traduções nos text views
     */
    private fun setLabels() {
        with(binding){
            act005OpcTvLoadingLogo.text = hmAux_Trans.get("drawer_loading_lbl")
            act005OpcTvPendencies.text = hmAux_Trans.get("lbl_unfinished_data")
            act005OpcTvHistoric.text = hmAux_Trans.get("drawer_historic_lbl")
            act005OpcTvEnableNfc.text = hmAux_Trans.get("toolbar_enable_nfc")
            act005OpcTvDisableNfc.text = hmAux_Trans.get("toolbar_cancel_nfc")
            act005OpcTvSupportData.text = hmAux_Trans.get("toolbar_support")
            act005OpcTvChangeCustomer.text = hmAux_Trans.get("lbl_change_customer")
            act005OpcTvLogout.text = hmAux_Trans.get("lbl_logout")
        }
    }

    /**
     * Metodo responsavel pelo setup de exibição das views q são opções condicionais.
     */
    private fun setSetupViews() {
        with(binding){
            val innerDelegate = delegate
            setCustomerLogoUI()
            //
            act005OpcTvEnableNfc.apply {
                visibility =
                        if(innerDelegate == null || !innerDelegate.showEnableNfcOption()){
                            View.GONE
                        }else{
                            View.VISIBLE
                        }
            }
            act005OpcTvDisableNfc.apply {
                visibility =
                        if(innerDelegate == null || !innerDelegate.showDisableNfcOption()){
                            View.GONE
                        }else{
                            View.VISIBLE
                        }
            }
            act005OpcTvChangeCustomer.apply {
                visibility =
                        if(innerDelegate == null || !innerDelegate.showChangeCustomerOption()){
                            View.GONE
                        }else{
                            View.VISIBLE
                        }
            }
        }
        //
        updatePendenceStatus()
    }

    /**
     * Metodo qe define o latou do que deve ser exibido no logo.
     * Se houver logo, exibe
     *  - Haverá logo se:
     *      1 - Customer com logo e logo baixado;
     *      2 - Customer sem logo, então exibisse o da namoa;
     * Se for null, exibe progress com texto
     *  - O retorno é null apenas se o customer possui logo definido mais ainda nao foi baixado.
     */
    private fun setCustomerLogoUI(){
        binding.act005OpcIvLogo.apply {
            val customerLogo = delegate?.getCustomerLogo()
            binding.act005OpcClLogo.visibility = if(customerLogo != null){
                View.GONE
            }else{
                View.VISIBLE
            }
            setImageBitmap(customerLogo)
        }
    }

    /**
     * Metodo publico camado pela act quando um ação pode refletir nas opções que devem ser exibidas.
     */
    fun revalidateOptionSetup(){
        setLabels()
        setSetupViews()
    }

    /**
     * Metodo que resgistra e desregistra o receiver do logo do cliente
     */
    private fun startNStopCustomerLogoReceiver(start: Boolean){
        val filter = IntentFilter()
        filter.addAction(ConstantBaseApp.BR_CUSTOMER_LOGO_ACTION)
        filter.addCategory(Intent.CATEGORY_DEFAULT)
        context?.let {
            if(start) {
                LocalBroadcastManager.getInstance(it).registerReceiver(logoReceiver, filter)
            }else{
                LocalBroadcastManager.getInstance(it).unregisterReceiver(logoReceiver)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        updatePendenceStatus()
    }

    override fun onStart() {
        super.onStart()
        //
        startNStopCustomerLogoReceiver(true)
    }

    override fun onStop() {
        super.onStop()
        //
        startNStopCustomerLogoReceiver(false)
    }

    /**
     * Metodo que revalida a exibição da informações de pendencias
     */
    private fun updatePendenceStatus() {
        binding.act005OpcTvPendencies.apply {
            val innerDelegate = delegate
            visibility =
                    if (innerDelegate == null || !innerDelegate.hasPendencies()) {
                        View.GONE
                    } else {
                        View.VISIBLE
                    }
        }
    }

    /**
     * Seta ações nas opções
     */
    private fun iniActions() {
        delegate?.let{ checkedDelegate->
            with(binding){

                act005OpcTvHistoric.setOnClickListener {
                    checkedDelegate.onHistoricClick()
                }
                act005OpcTvEnableNfc.setOnClickListener {
                    checkedDelegate.onEnableNfcClick()
                }
                act005OpcTvDisableNfc.setOnClickListener {
                    checkedDelegate.onDisableNfcClick()
                }
                act005OpcTvSupportData.setOnClickListener {
                    checkedDelegate.onSupportDataClick()
                }
                act005OpcTvChangeCustomer.setOnClickListener {
                    checkedDelegate.onChangeCustomerClick()
                }
                act005OpcTvLogout.setOnClickListener {
                    checkedDelegate.onLogoutClick()
                }
            }
        }
    }

    /**
     * LUCHE - 04/05/2021
     * Criado class broadcast receiver para interceptar "notificação" de que iconde foi baixado.
     */
    inner class CustomerLogoReceiver : BroadcastReceiver(){
        override fun onReceive(context: Context?, intent: Intent?) {
            setCustomerLogoUI()
        }
    }
}